onnxruntime/contrib_ops/cpu/gather_nd.h

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/common/common.h"
+#include "core/framework/op_kernel.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+class GatherNDBase
+{
+protected:
+
+  struct Prepare {
+    const uint8_t*        input_base;
+    const std::string*    input_str_base;
+    uint8_t*              output_base;
+    std::string*          output_str_base;
+    uint64_t              bytes_to_copy;
+    uint64_t              element_bytes;
+    uint64_t              element_to_copy;
+    std::vector<uint64_t> element_offsets;
+
+    Prepare(): input_base      (nullptr),
+               input_str_base  (nullptr),
+               output_base     (nullptr),
+               output_str_base (nullptr),
+               bytes_to_copy   (0),
+               element_bytes   (0),
+               element_to_copy (0),
+               element_offsets (0) {}
+  }; // struct Prepare
+
+  template<typename Tind>
+  Status PrepareForCompute(OpKernelContext* context, Prepare& p) const;
+}; // class GatherNDBase
+
+template<typename Tind>
+class GatherNDString final : public OpKernel, protected GatherNDBase {
```

Comment:
>GatherNDString [](start = 6, length = 14)

no need to have 2 separate class, we can get the input data type.
[https://github.com/Microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cpu/tensor/gather.cc#L77](https://github.com/Microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cpu/tensor/gather.cc#L77)

Reply:
Yes I see that from Gather implementation.
In Gather, the implementation keeps a BOOL - "is_string_type", then in a for-for loop, it judges every single time whether the input type is a string.
By template, we could move this efforts to compile time.

Reply:
the reason we want to get rid of template is to keep the binary size as small as possible. I think it won't impact much put the if inside the loop.

---
In reply to: [241870901](https://github.com/Microsoft/onnxruntime/pull/170#discussion_r241870901) [](ancestors = 241870901)

Reply:
Got it, now they are removed.